### PR TITLE
feat: Implement POLARS_MAX_MEMORY_MIB

### LINKED ISSUE #9892: Custom memory limits
### Problem description

A feature like duckDB's ```SET memory_limit='{x}GB';``` would be useful for those of us working in the PBS Pro world, since polars currently does not respect the PBS memory limits. `ulimit` or the standard library `resource` is a partial workaround since it allows error handling on OOM instead of a PBS job kill, but unless I am missing something (very possible), polars will always try to allocate to memory it sees as available at the system level.

A simple config variable would be a POLARS_MAX_THREADS equivalent, say POLARS_MAX_MEMORY.

### PR DESCRIPTION
**Pull request by jonashaag**
Fixes #9892

The code seem to work in that it overrides the memory reported by sysinfo, but I'm unable to have Polars actually comply with the limits I'm setting.

### CODE REVIEW

File: crates/polars-pipe/src/executors/sinks/memory.rs [hunk 1]
```diff
@@ -25,25 +26,30 @@ impl MemTracker {
...
+        out.available_at_start = out.available_mem.load(Ordering::Relaxed);
         out
     }
 
     /// This shouldn't be called often as this is expensive.
     fn refresh_memory(&self) {
+        let free = MEMINFO.free() as usize;
         self.available_mem
-            .store(MEMINFO.free() as usize, Ordering::Relaxed);
+            .store(self.available_mem_max.unwrap_or(free).min(free), Ordering::Relaxed);
```
Comment (cesarkawakami) [inline crates/polars-pipe/src/executors/sinks/memory.rs:52]: Shouldn't this be something like `self.available_mem_max - self.used_mem`, though I realize `used_mem` would also require tracking?

Comment (jonashaag) [reply crates/polars-pipe/src/executors/sinks/memory.rs:52]: Why not use `.free()` if we have it in the sysinfo API?

**Comment by stinodego** [comment]
Thanks for taking the initiative on this one.

> I'm unable to have Polars actually comply with the limits I'm setting

That means this is not ready to be merged, right? I'm putting this on draft for now. If you're looking for help on this, please tag a code owner.

**Comment by jonashaag** [comment]
> please tag a code owner.

OK, tagging @ritchie46 @orlp!

**Comment by shenker** [comment]
Any progress on this? Would greatly appreciate this (been running into memory limits running on an academic SLURM cluster)! Happy to test if/when that would be helpful.

**Comment by jonashaag** [comment]
It's not working and I don't know why. Please feel free to give it a try and help understanding why it's not working.

**Comment by mkleinbort-ic** [comment]
Yes, this feature would be great. In my case I have multiple processes running polars operations at the same time on the same server, and when the steps "sync" the memory usage is over 9000. It also makes it hard to test the speed-vs-resources tradeoff (e.g. could I be running the same code on a smaller instance?)

**Comment by aecorn** [comment]
Highly interested in this. Google Cloud Run limits memory to 32Gb, and we have bigger datasets than will fit in that memory that we want to "push through and clean up columns on". Could this help with something like setting a "known upper limit" for MEM usage by Polars, so it doesnt crash the containers on OOM?

**Comment by jonashaag** [comment]
Does Google Cloud Run properly set the limits via cgroups?

**Comment by aecorn** [comment]
> Google Cloud Run properly set the limits via cgroups?

[Seems like it.](https://cloud.google.com/run/docs/about-execution-environments)

**Comment by jonashaag** [comment]
Then memory reporting should work fine with recent versions of Polars (I implemented cgroups support in the rs-sysinfo crate that's used by Polars)

**Comment by stinodego** [comment]
I'll close this as there has been no progress here in a while. Please open a new PR if you have something working!

**Comment by ritchie46** [comment]
@stinodego was their something that should still be done on OPs side? I think we still had to get to this one. Evwn though we cannot get to it soon enough there might be something of value in it. I haven't had the time to look yet.

**Comment by stinodego** [comment]
We definitely still want this functionality, but @jonashaag mentioned this code wasn't working yet but didn't know why.

Since there wasn't any progress here in a while I figured I'd close it, maybe someone else could pick it up. But if you can help get this code working then that's even better!

**Comment by ritchie46** [comment]
Ok, then I understand. :)

**Comment by devYonz** [comment]
What is the best way to track this effort? Having some sort of memory limit feels like a must.

**Comment by stinodego** [comment]
> What is the best way to track this effort? Having some sort of memory limit feels like a must.

You can subscribe to the linked issue.

The _best_ way is to implement this functionality and open a PR :)

**Comment by mishpat** [comment]
Any chance that #15798 helps with this? Looks like polars' version of sysinfo has been bumped, so while the issue surely isn't the same the sysinfo bump may have sorted out the problem.
This is fishing because I don't have the knowledge to work on this in a reasonable timeframe, but if it could get working that would be great.

**Comment by antonio-yuen-zocdoc** [comment]
> Any chance that #15798 helps with this? Looks like polars' version of sysinfo has been bumped, so while the issue surely isn't the same the sysinfo bump may have sorted out the problem. This is fishing because I don't have the knowledge to work on this in a reasonable timeframe, but if it could get working that would be great.

I'm not so sure that it does. That was introduced on Polars 0.40.0 and I'm on polars 1.5.0 running on ECS Fargate instances. I see my clusters get killed from exceeding memory limits all the time. :(

**Comment by mishpat** [comment]
This issue is likely dead until the new streaming engine is finished. Hoping that the new implementation makes some sort of custom memory limits possible.

**Pull request closed without merging**